File: src/main/java/myblog/blog/base/log/LogTracer.java
package myblog.blog.base.log;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.context.request.RequestContextHolder;
import org.springframework.web.context.request.ServletRequestAttributes;

import javax.servlet.http.HttpServletRequest;

@Slf4j
@Component
public class LogTracer {

    private static final String START_PREFIX = "-->";
    private static final String COMPLETE_PREFIX = "<--";
    private static final String EX_PREFIX = "<X-";

    private ThreadLocal<TraceId> traceIdHolder = new ThreadLocal<>();

    public TraceStatusVO begin(String message, String args){
        syncTraceId();
        TraceId traceId = traceIdHolder.get();
        Long startTimeMs = System.currentTimeMillis();
        if(traceId.isFirstLevel()) {
            log.info("------------------------------"+traceId.getId()+"'s transaction start------------------------------");
        }
        log.info("[{}] {}{} ,args = {}",traceId.getId(), addSpace(START_PREFIX, traceId.getLevel()), message, args);
        return new TraceStatusVO(traceId, startTimeMs, message);
    }

    private void syncTraceId() {
        TraceId traceId = traceIdHolder.get();
        if (traceId == null) {
            HttpServletRequest request = ((ServletRequestAttributes) RequestContextHolder.currentRequestAttributes()).getRequest();
            String clientIP = IpGetHelper.getClientIP(request);
            traceIdHolder.set(new TraceId(clientIP));
        } else {
            traceIdHolder.set(traceId.createNextId());
        }
    }

    public void end(TraceStatusVO traceStatusVO){
        complete(traceStatusVO, null);
    }

    public void handleException(TraceStatusVO traceStatusVO, Exception ex){
        complete(traceStatusVO, ex);
    }

    private void complete(TraceStatusVO traceStatusVO, Exception ex) {
        Long stopTimeMs = System.currentTimeMillis();
        Long resultTimeMs = stopTimeMs - traceStatusVO.getStartTimesMs();
        TraceId traceId = traceStatusVO.getTraceId();
        if(ex == null){
            log.info("[{}] {}{} time = {}ms", traceId.getId(), addSpace(COMPLETE_PREFIX, traceId.getLevel()),
                    traceStatusVO.getMessage(), resultTimeMs);
        } else {
            log.info("[{}] {} {} time = {}ms ex={}", traceId.getId(), addSpace(EX_PREFIX, traceId.getLevel()),
                    traceStatusVO.getMessage(), resultTimeMs, ex.toString());
        }
        if(traceStatusVO.getTraceId().isFirstLevel()) {
            log.info("-------------------------------"+traceId.getId()+"'s transaction end/"+resultTimeMs+"ms-------------------------");
        }
        releaseTraceId();
    }

    private void releaseTraceId() {
        TraceId traceId = traceIdHolder.get();
        if (traceId.isFirstLevel()) {
            traceIdHolder.remove(); //destroy
        } else {
            traceIdHolder.set(traceId.createPrevId());
        }
    }

    private String addSpace(String prefix, int level) {
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i< level; i++){
            sb.append((i==level-1) ? "|" + prefix : "|   ");
        }
        return sb.toString();
    }
}

File: libs/global-logging/src/main/kotlin/kr/co/jiniaslog/shared/log/LogTracer.kt
package kr.co.jiniaslog.shared.log

import jakarta.servlet.http.HttpServletRequest
import org.springframework.stereotype.Component
import org.springframework.web.context.request.RequestContextHolder
import org.springframework.web.context.request.ServletRequestAttributes
import java.time.LocalDateTime

private val log = mu.KotlinLogging.logger {}

@Component
class LogTracer {
    private val traceIdHolder = ThreadLocal<TraceId>()
    private val logBuffer = ThreadLocal<MutableList<String>>()

    fun begin(
        message: String,
        args: String,
        returnType: String,
    ): TraceStatus {
        syncTraceId()
        val traceId = traceIdHolder.get()
        val startTimeMs = System.currentTimeMillis()
        if (traceId.isFirstLevel) {
            addToLogBuffer("------------------------------" + traceId.id + "'s transaction start------------------------------")
        }
        addToLogBuffer(
            "[${traceId.id}] ${
                addSpace(
                    START_PREFIX,
                    traceId.level,
                )
            }$message ,args = $args, returnType = $returnType",
        )
        return TraceStatus(traceId, startTimeMs, message)
    }

    private fun syncTraceId() {
        val traceId = traceIdHolder.get()
        val log = logBuffer.get()
        if (traceId == null) {
            try {
                val request: HttpServletRequest =
                    (RequestContextHolder.currentRequestAttributes() as ServletRequestAttributes).request
                val clientIP = IpGetHelper.getClientIP(request)
                traceIdHolder.set(TraceId.create(clientIP))
                logBuffer.set(mutableListOf())
            } catch (e: Exception) {
                traceIdHolder.set(TraceId.create("not found ip"))
                logBuffer.set(mutableListOf())
            }
        } else {
            traceIdHolder.set(traceId.createNextId())
        }
    }

    fun end(traceStatusVO: TraceStatus) {
        complete(traceStatusVO, null)
    }

    fun handleException(
        traceStatusVO: TraceStatus,
        ex: Exception?,
    ) {
        complete(traceStatusVO, ex)
    }

    private fun complete(
        traceStatus: TraceStatus,
        ex: Exception?,
    ) {
        val stopTimeMs = System.currentTimeMillis()
        val resultTimeMs: Long = stopTimeMs - traceStatus.startTimesMs
        val traceId: TraceId = traceStatus.traceId
        if (ex == null) {
            addToLogBuffer(
                "[${traceId.id}] ${addSpace(COMPLETE_PREFIX, traceId.level)}${traceStatus.message} time = ${resultTimeMs}ms",
            )
        } else {
            addToLogBuffer(
                "[${traceId.id}] ${
                    addSpace(
                        EX_PREFIX,
                        traceId.level,
                    )
                } ${traceStatus.message} time = ${resultTimeMs}ms ex = $ex",
            )
        }
        if (traceStatus.traceId.isFirstLevel) {
            addToLogBuffer(
                "-------------------------------" + traceId.id + "'s transaction end/" + resultTimeMs + "ms-------------------------",
            )
        }
        releaseAndFlushIfNeeded()
    }

    private fun releaseAndFlushIfNeeded() {
        val traceId = traceIdHolder.get()
        if (traceId.isFirstLevel) {
            flushLogs()
            traceIdHolder.remove()
            logBuffer.remove()
        } else {
            traceIdHolder.set(traceId.createPrevId())
        }
    }

    private fun addSpace(
        prefix: String,
        level: Int,
    ): String {
        val sb = StringBuilder()
        for (i in 0 until level) {
            sb.append(if ((i == level - 1)) "|$prefix" else "|   ")
        }
        return sb.toString()
    }

    fun flushLogs() {
        logBuffer.get().reduce { first, second ->
            first + "\n" + second
        }.also {
            log.info("\n$it")
        }
        logBuffer.remove()
    }

    private fun addToLogBuffer(message: String) {
        logBuffer.get().add(LocalDateTime.now().toString() + " : " + message)
    }

    companion object {
        private const val START_PREFIX = "-->"
        private const val COMPLETE_PREFIX = "<--"
        private const val EX_PREFIX = "<X-"
    }
}
